File: scripts/verification_worker.py
#!/usr/bin/env python3

import json
import os
import time
import urllib.request
import urllib.error
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]

ENV_PATH = (
    ROOT /
    "apps/web/.env.local"
)

PLATFORM_URL = (
    os.environ.get(
        "AGENTBOUNTY_PLATFORM_URL",
        "http://localhost:3000",
    )
    .strip()
    .rstrip("/")
)

INTERVAL_SECONDS = int(
    os.environ.get(
        "AGENTBOUNTY_VERIFICATION_INTERVAL_SECONDS",
        "15",
    )
)


def load_internal_key():
    env_value = (
        os.environ.get(
            "AGENTBOUNTY_INTERNAL_KEY",
            ""
        )
        .strip()
    )

    if env_value:
        return env_value

    if ENV_PATH.exists():
        for raw in (
            ENV_PATH
            .read_text()
            .splitlines()
        ):
            line = raw.strip()

            if (
                not line
                or line.startswith("#")
                or "=" not in line
            ):
                continue

            key, value = line.split(
                "=",
                1
            )

            if (
                key ==
                "AGENTBOUNTY_INTERNAL_KEY"
            ):
                value = value.strip()

                if value:
                    return value

    raise RuntimeError(
        "AGENTBOUNTY_INTERNAL_KEY missing"
    )


def run_once():
    internal_key = (
        load_internal_key()
    )

    request = urllib.request.Request(
        PLATFORM_URL +
        "/api/v1/system/verification-queue",
        method="POST",
        headers={
            "x-internal-key":
                internal_key,

            "Content-Type":
                "application/json",
        },
    )

    with urllib.request.urlopen(
        request,
        timeout=60,
    ) as response:
        return json.load(
            response
        )


def main():
    print(
        "AgentBounty Verification Worker",
        flush=True,
    )

    print(
        f"Platform: {PLATFORM_URL}",
        flush=True,
    )

    print(
        f"Polling every "
        f"{INTERVAL_SECONDS}s",
        flush=True,
    )

    while True:
        try:
            data = run_once()

            scanned = data.get(
                "scanned",
                0
            )

            if scanned:
                print(
                    f"[verification] "
                    f"scanned "
                    f"{scanned} task(s)",
                    flush=True,
                )

                for item in data.get(
                    "results",
                    []
                ):
                    task_id = item.get(
                        "taskId",
                        "?"
                    )

                    result = (
                        item.get(
                            "verificationStatus"
                        )
                        or item.get(
                            "error"
                        )
                        or "unknown"
                    )

                    status = item.get(
                        "status",
                        ""
                    )

                    suffix = (
                        f" -> {status}"
                        if status
                        else ""
                    )

                    print(
                        f"  {task_id}: "
                        f"{result}"
                        f"{suffix}",
                        flush=True,
                    )

        except KeyboardInterrupt:
            print(
                "Verification worker stopped.",
                flush=True,
            )
            break

        except urllib.error.HTTPError as exc:
            body = (
                exc.read()
                .decode(
                    "utf-8",
                    errors="replace"
                )
            )

            print(
                f"[verification] "
                f"HTTP {exc.code}: "
                f"{body}",
                flush=True,
            )

        except Exception as exc:
            print(
                f"[verification] "
                f"error: {exc}",
                flush=True,
            )

        time.sleep(
            INTERVAL_SECONDS
        )


if __name__ == "__main__":
    main()

File: scripts/verification_worker.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename =
  fileURLToPath(import.meta.url);

const __dirname =
  path.dirname(__filename);

const ROOT =
  path.resolve(__dirname, "..");

const ENV_PATH =
  path.join(
    ROOT,
    "apps/web/.env.local"
  );

const PLATFORM_URL =
  (
    process.env
      .AGENTBOUNTY_PLATFORM_URL
    || "http://localhost:3000"
  )
    .trim()
    .replace(/\/+$/, "");

const INTERVAL_SECONDS =
  Number(
    process.env
      .AGENTBOUNTY_VERIFICATION_INTERVAL_SECONDS
    || "15"
  );

function loadInternalKey() {
  const envValue =
    (
      process.env
        .AGENTBOUNTY_INTERNAL_KEY
      || ""
    ).trim();

  if (envValue) {
    return envValue;
  }

  if (
    fs.existsSync(
      ENV_PATH
    )
  ) {
    const lines =
      fs.readFileSync(
        ENV_PATH,
        "utf8"
      )
      .split(/\r?\n/);

    for (
      const raw
      of lines
    ) {
      const line =
        raw.trim();

      if (
        !line
        || line.startsWith("#")
        || !line.includes("=")
      ) {
        continue;
      }

      const index =
        line.indexOf("=");

      const key =
        line
          .slice(0, index)
          .trim();

      const value =
        line
          .slice(index + 1)
          .trim();

      if (
        key ===
          "AGENTBOUNTY_INTERNAL_KEY"
        && value
      ) {
        return value;
      }
    }
  }

  throw new Error(
    "AGENTBOUNTY_INTERNAL_KEY missing"
  );
}

async function runOnce() {
  const internalKey =
    loadInternalKey();

  const response =
    await fetch(
      `${PLATFORM_URL}/api/v1/system/verification-queue`,
      {
        method: "POST",

        headers: {
          "x-internal-key":
            internalKey,

          "content-type":
            "application/json",
        },

        signal:
          AbortSignal.timeout(
            60_000
          ),
      }
    );

  const text =
    await response.text();

  if (
    !response.ok
  ) {
    throw new Error(
      `HTTP ${response.status}: ${text}`
    );
  }

  return text
    ? JSON.parse(text)
    : {};
}

async function main() {
  console.log(
    "AgentBounty Verification Worker"
  );

  console.log(
    `Platform: ${PLATFORM_URL}`
  );

  console.log(
    `Polling every ${INTERVAL_SECONDS}s`
  );

  while (true) {
    try {
      const data =
        await runOnce();

      const scanned =
        data.scanned ?? 0;

      if (scanned) {
        console.log(
          `[verification] scanned ${scanned} task(s)`
        );

        for (
          const item
          of data.results ?? []
        ) {
          const taskId =
            item.taskId ?? "?";

          const result =
            item.verificationStatus
            ?? item.error
            ?? "unknown";

          const suffix =
            item.status
              ? ` -> ${item.status}`
              : "";

          console.log(
            `  ${taskId}: ${result}${suffix}`
          );
        }
      }
    } catch (error) {
      console.error(
        "[verification] error:",
        error instanceof Error
          ? error.message
          : String(error)
      );
    }

    await new Promise(
      resolve =>
        setTimeout(
          resolve,
          INTERVAL_SECONDS * 1000
        )
    );
  }
}

main().catch(error => {
  console.error(error);
  process.exit(1);
});
